Boundary restrictable InterfaceUserObject (and postprocessor) family

arovinelli opened the pull request:
#12902 was close by mistake I reopen the same PR here

- The aim of this new class is to provide an **easy to use** user object to make computation across an interface.

- As mentioned previously this class is not. The`InterfaceUserObject` is boundary restricted and allows to make computation on variable across an interface.  Some addition has been done to `FEProblemBase` to allow this new class to properly work and `ComputeUserObjectThread`.
Note that, because we are still waiting on  #12101 this user object is not able to work on material properties at the moment. In fact @lindsayad request #12079 to have a PR without modification tot he material portion of moose. The workaround for this is to convert material property in aux variables and then use the variable as input for `InterfaceUserObject`

- in general the `InterfaceUserObject` is helpful for all that situations in which a UO is needed to do computation across an Interface where the `InterfaceKernel` is used.
continuation of #11660
 ref #11647 


@lindsayad @YaqiWang  @vincentlaboure have helped during development found bags etc, so they already know what this PR is about

Commit:
for PR, issue ref #11647

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/291772/) on 27f5d1f wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12902/site)

This comment will be updated on new commits.

Commit:
delete some commented line

arovinelli commented:
@permcody @YaqiWang @lindsayad The code is ok, I just have a question about the failure of distributed mesh. In fact on my machine the test passes but I receive a warning 
```
Warning:  This MeshOutput subclass only supports meshes which have been serialized!
```

what could it be?

YaqiWang commented:
That warning is benign. There must be something else going on.

permcody commented:
This isn't an error, just a warning. It's telling you that you have a distributed mesh, but you are writing it to a serial (e.g. exodus) format (so it must be serialized, which is occurring because that's the way we do it in MOOSE). Normally if you have a distributed mesh, you'd write a distributed format (e.g. nemesis).

YaqiWang commented:
I see the error in `diffusivity_1` like
```
diffusivity_1          rel diff:  1.0000000e+01 ~  4.0000000e+00 = 6.00000e-01 (block 0, elmt 1)
```
It does not make sense to me because the variable is set by the aux kernel, on block 0 first time step it should be just 10. Could possible somewhere else the variable be written?

arovinelli commented:
@YaqiWang and @permcody thanks for the reply.
I'm still puzzled about failure in the test machine. It's complaining about elemental variables coming from boundary/block restricted auxkernels, even simple variable coming out from a MaterialRealAux.
The problem is I can't replicate this behavior on my machine...

YaqiWang commented:
Even with the exact same command?
```
mpiexec -n 4 /opt/civet/build_0/moose/test/moose_test-opt -i interface_user_object_QP.i --distributed-mesh --error --error-unused --error-override --timing Outputs/perf_graph=true --no-gdb-backtrace
```

arovinelli commented:
@YaqiWang , I even tried changing the number of CPU but nothing...
The user object has it also declared as constant reference...

YaqiWang commented:
This could be related element numbering in distributed mesh. Can you get rid of the UO stuff in the input and just create a simple test with the material and aux kernel? Add one commit on top of this to see if you can observe the same behavior.

Commit:
test element numbering

arovinelli commented:
@YaqiWang done.. let's wait and see

arovinelli commented:
@YaqiWang it does fail too the simple test, please check test_element_numbering inside userobject/interface_user_object

YaqiWang commented:
This might be caused by the AssignElementID mesh modifier. Can you replace it with BoundingBox one to see if we get the same? I am outside cannot check the code for now.

Commit:
changed input file to avoid using addElementSubdomainID in fovour of subdomainBoundingBox

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit c946e65a245948f384835f05a88e6b9a13a1a090
+Subproject commit 6c55e2f9af9616953b98658e909ab9be805ae751
```
**Caution!** This contains a submodule update

arovinelli commented:
@cody and @YaqiWang if I register a class (e.g. InterfaceUO ) under MooseTestApp and write documentation, than i can't add the !syntax parameters and stuff like that becasue it fails tests. What should I do with all the example and classes that I made for the tests and examples? I jsut put a description without putting the following or there is a way around this (beside registering in MooseApp)?

```
!syntax parameters /UserObjects/InterfaceUO
!syntax inputs /UserObjects/InterfaceUO
!syntax children /UserObjects/InterfaceUO
```

permcody commented:
We currently aren't documenting the Test objects. We are focusing on documenting the Framework and modules. Just omit those .md files for now.

arovinelli commented:
@cody, thanks for the reply. If you are ok with it, Ill leave the md file of the tests classes as I already wrote them, but I'll deleted the lines making it fail.

permcody commented:
@aeslaughter - Any comments on the Test app .md files?

Commit:
update some stuff and doc

arovinelli commented on framework/src/userobject/InterfaceUserObject.C:
```diff
@@ -0,0 +1,61 @@
+    _q_point(_assembly.qPointsFace()),
+    _qrule(_assembly.qRuleFace()),
+    _JxW(_assembly.JxWFace()),
+    _JxW_neighbor(_assembly.JxWNeighbor()),
```
I would like to give access to this user object the `_JxW_neighbor`. However if I use _JxW_neighbor[qp] in exectue I receive an assertion error:

```
Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 0 size: 0)
at /home/arovinelli/projects/moose/framework/build/header_symlinks/MooseArray.h, line 268
Detaching after fork from child process 6128.
Stack frames: 15
0: libMesh::print_trace(std::ostream&)
1: MooseArray<double>::operator[](unsigned int) const
2: InterfaceUO::execute()
3: ComputeUserObjectsThread::onInterface(libMesh::Elem const*, unsigned int, short)
4: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
5: void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeUserObjectsThread>(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, ComputeUserObjectsThread&)
6: FEProblemBase::computeUserObjects(MooseEnumItem const&, Moose::AuxGroup const&)
7: FEProblemBase::execute(MooseEnumItem const&)
8: FEProblemBase::initialSetup()
9: Steady::init()
10: MooseApp::executeExecutioner()
11: MooseApp::run()
12: /home/arovinelli/projects/moose/test/tests/userobjects/interface_user_object/../../../moose_test-dbg() [0x40e818]
13: __libc_start_main
14: /home/arovinelli/projects/moose/test/tests/userobjects/interface_user_object/../../../moose_test-dbg() [0x40e529]
[0] /home/arovinelli/projects/moose/framework/build/header_symlinks/MooseArray.h, line 268, compiled Feb 15 2019 at 08:30:52
```

Am I missing something to reinit something in the loops  or simply  this not supposed to be used?

arovinelli replied:
@jwpeterson any idea about this?

arovinelli replied:
For now, I removed access to _JxW_neighbor. This implies small deformations for a disconnect mesh (e.g when using `BreakMeshByBlock`).

Commit:
fixed test file...

Commit:
supposvly fix the wrong test

arovinelli commented:
@permcody @lindsayad @YaqiWang this is ready for review

moosebuild commented:
Job [Modules debug PETSc submodule](https://civet.inl.gov/job/296539/) on a84b9db : invalidated by @arovinelli

machine went out of space

arovinelli commented:
@permcody ping

YaqiWang commented:
@arovinelli MOOSE guys are at SIAM meeting this week so they may not be very responsive. @vincentlaboure do you want to take a look on this?

arovinelli commented:
@YaqiWang Thanks for letting me know. Of course in the meantime the MOOSE guys are at SIAM, whoever has an use for this, like @vincentlaboure can start checking and commenting

vincentlaboure commented:
@arovinelli Sorry, I just saw this. I can take a look but the MOOSE guys will have to go after me

vincentlaboure commented on framework/doc/content/source/postprocessors/InterfaceIntegralVariableAveragePostprocessor.md:
```diff
@@ -0,0 +1,21 @@
+# InterfaceIntegralVariableAveragePostprocessor
+
+## IMPORTANT NOTE
+*** Any derived class of InterfaceUserObject does not support yet getMaterialProperty, please use auxvariables ***
```
I didn't follow all the history of your implementation. Can you summarize why `getMaterialProperty` can't be called in this UO? I have a class deriving from it and calling `getMaterialProperty` and `getNeighborMaterialProperty`, and at first glance I didn't see any problem with that.

arovinelli replied:
@vincentlaboure `getMaterialProperty` should trigger the calculation of material property. However, in the loops `ComputeMaterialThreads:OnInterface` is not defined yet therefore there is no guarantee you will get the most update material property when executing the UO. You should not see differences if you execute the UO with `execute_on=TIMESTEP_END` but you should see bad results if you use `execute_on=LINEAR`.  This is the reason why you should not use `getMaterialProperty` from an `InterfaceUO` until PR #12101 is completed and merged. The workaround to this is to dump MP into `AuxVars` and use them inside the `InterfaceUO`.
You should be able to check this bad behaviour easily. let me know if you try and it works :)

vincentlaboure commented on framework/doc/content/source/postprocessors/InterfaceIntegralVariableAveragePostprocessor.md:
```diff
@@ -0,0 +1,21 @@
+## Description
+!syntax description /Postprocessors/InterfaceIntegralVariableAveragePostprocessor
+
+This post-processor is used to compute an average value across an interface. The type of average to compute is selected by changing the `average_type` parameter. If the parameter `average_type` is omitted it default on computing the average value between the two input variables across the interface.
```
it default on computing -> it __defaults to__ computing

vincentlaboure commented on framework/doc/content/source/postprocessors/InterfaceIntegralVariableAveragePostprocessor.md:
```diff
@@ -0,0 +1,21 @@
+!syntax description /Postprocessors/InterfaceIntegralVariableAveragePostprocessor
+
+This post-processor is used to compute an average value across an interface. The type of average to compute is selected by changing the `average_type` parameter. If the parameter `average_type` is omitted it default on computing the average value between the two input variables across the interface.
+The kind of average to compute are describe in more details in  [InterfaceAverageTools](/InterfaceAverageTools.md).
```
The kind of average to compute are describe -> The __various types__ of __averages__ are __described__...

vincentlaboure commented on framework/doc/content/source/utils/InterfaceAverageTools.md:
```diff
@@ -0,0 +1,17 @@
+
+Basic utility file to compute a single quantity given values on both side of an interface. The value to compute is selected via the input parameter
+average_type. This function is not registered and as such should only be used for simplifying the development of other objects.
+Amn example of an object using this utility is  [InterfaceIntegralVariableAveragePostprocessor](/InterfaceIntegralVariableAveragePostprocessor.md).
```
Amn example -> An example

vincentlaboure commented on framework/include/postprocessors/InterfaceIntegralVariablePostprocessor.h:
```diff
@@ -0,0 +1,47 @@
+  /// Holds the solution gradient at the current quadrature points on the
+  /// neighbor side
+  const VariableGradient & _grad_u_neighbor;
+  // virtual Real computeQpIntegral() override = 0;
```
Remove commented line?

vincentlaboure commented on framework/include/userobject/InterfaceAverageUserObject.h:
```diff
@@ -0,0 +1,35 @@
+protected:
+  /// the average type to be computed across the interface
+  const MooseEnum _average_type;
+  virtual Real ComputeAverageType(const Real /*value_master*/, const Real /*value_slave*/);
```
This method may need to be renamed `computeAverageType` to follow standard naming conventions

arovinelli replied:
done

vincentlaboure commented on framework/include/utils/InterfaceAverageTools.h:
```diff
@@ -0,0 +1,37 @@
+MooseEnum InterfaceAverageOptions();
+
+/*
+ * Return scalar quantity across an interface o based on the user specified
```
Extra 'o' (remove or replace)

vincentlaboure commented on framework/include/utils/InterfaceAverageTools.h:
```diff
@@ -0,0 +1,37 @@
+
+/*
+ * Return scalar quantity across an interface o based on the user specified
+ * _average_type. First paramter is the vaerage type, the second  the value
```
vaerage -> average
paramter -> parameter
remove one space after 'second'

vincentlaboure commented on framework/src/postprocessors/InterfaceIntegralPostprocessor.C:
```diff
@@ -0,0 +1,64 @@
+{
+  InputParameters params = validParams<InterfacePostprocessor>();
+  params.addClassDescription(
+      "Postprocessor computing the intgral of a quantity on an interface. Different type of "
```
intgral -> integral
Different type -> Different types
remove one space after 'overriding'

lindsayad replied:
It looks like this wasn't actually resolved

vincentlaboure commented on framework/src/postprocessors/InterfaceIntegralVariablePostprocessor.C:
```diff
@@ -0,0 +1,45 @@
+  params.addRequiredCoupledVar("variable",
+                               "The name of the variable on the master side of the interface");
+  params.addCoupledVar("neighbor_variable",
+                       "The name of the variable on the slave side of the itnerface. By default "
```
itnerface -> interface

vincentlaboure commented on framework/src/postprocessors/InterfaceIntegralVariablePostprocessor.C:
```diff
@@ -0,0 +1,45 @@
+  params.addCoupledVar("neighbor_variable",
+                       "The name of the variable on the slave side of the itnerface. By default "
+                       "the same variable name is used for the slave side");
+  params.addClassDescription("Add access to variables and their gradinet on an interface.");
```
gradinet -> gradient

vincentlaboure commented on framework/src/userobject/InterfaceAverageUserObject.C:
```diff
@@ -0,0 +1,34 @@
+  params.addParam<MooseEnum>(
+      "average_type", InterfaceAverageTools::InterfaceAverageOptions(), "Type of scalar output");
+  params.addClassDescription(
+      "Special subclass of interface userobject givng access to interface average utilities.");
```
givng -> giving

vincentlaboure commented on framework/src/userobject/InterfaceUserObject.C:
```diff
@@ -0,0 +1,61 @@
+const Real &
+InterfaceUserObject::getNeighborElemVolume()
+{
+
```
remove empty line

vincentlaboure commented on framework/src/utils/InterfaceAverageTools.C:
```diff
@@ -0,0 +1,59 @@
+      result = value_slave;
+      break;
+    default:
+      mooseError("InterfaceIntegralMaterialPropertyPostprocessor: the supplied intergal "
```
intergal -> integral
Avaialble -> Available

vincentlaboure commented on test/doc/content/source/auxkernels/InterfaceUserObjectQpAux.md:
```diff
@@ -0,0 +1,11 @@
+
+## Description
+
+InterfaceUserObjectQpAux is an AuxKernel Used to collect average value collected on the interface via [InterfaceUO_QP](/InterfaceUO_QP.md) and save them into an aux variable at each quadrature point.
```
Used -> used

vincentlaboure commented on test/doc/content/source/postprocesors/InterfaceUOPPS.md:
```diff
@@ -0,0 +1,14 @@
+!syntax description /Postprocessors/InterfaceUOPPS
+
+InterfaceUOPPS is a postprocessor for [InterfaceUO](/InterfaceUO.md).
+It create a global variable from the value computed in the linked userobject.
```
create -> creates

vincentlaboure commented on test/doc/content/source/userobjects/InterfaceUO.md:
```diff
@@ -0,0 +1,13 @@
+## Description
+!syntax description /UserObjects/InterfaceUO
+
+InterfaceUO is a user object computing an average value across an interface. The kind of average value are the one available in  [InterfaceAverageTools](/InterfaceAverageTools.md).
```
The kind of average value are the one available in -> The **various types** of **averages** are **described** in...

vincentlaboure commented on test/doc/content/source/userobjects/InterfaceUO_QP.md:
```diff
@@ -0,0 +1,16 @@
+## Description
+!syntax description /UserObjects/InterfaceUO_QP
+
+InterfaceUO_QP is a user object computing and storing average values across an interface for each quadrature. The kind of average value are the one available in [InterfaceAverageTools](/InterfaceAverageTools.md).
```
for each quadrature -> __at__ each quadrature __point__
The kind of average value are the one available in -> The **various types** of **averages** are **described** in...

vincentlaboure commented on test/include/auxkernels/InterfaceUserObjectQpAux.h:
```diff
@@ -0,0 +1,39 @@
+InputParameters validParams<InterfaceUserObjectQpAux>();
+
+/**
+ * Self auxiliary value
```
This should be changed

vincentlaboure commented on test/include/postprocessors/InterfaceUOPPS.h:
```diff
@@ -0,0 +1,42 @@
+
+  virtual void initialize() override;
+  virtual void execute() override;
+  // virtual void finalize() override;
```
remove commented line

vincentlaboure commented on test/include/userobjects/InterfaceUO.h:
```diff
@@ -0,0 +1,44 @@
+InputParameters validParams<InterfaceAverageUserObject>();
+
+/**
+ *
```
Empty comment lines

vincentlaboure commented on test/include/userobjects/InterfaceUO_QP.h:
```diff
@@ -0,0 +1,44 @@
+InputParameters validParams<InterfaceUO_QP>();
+
+/**
+ *
```
Empty comment lines

vincentlaboure commented on test/src/auxkernels/InterfaceUserObjectQpAux.C:
```diff
@@ -0,0 +1,34 @@
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addRequiredParam<UserObjectName>("interface_uo_name",
+                                          "the name of the interface user object to use");
```
Upper case to start the doc string

vincentlaboure commented on test/src/postprocessors/InterfaceUOPPS.C:
```diff
@@ -0,0 +1,49 @@
+  InputParameters params = validParams<GeneralPostprocessor>();
+  params.addRequiredParam<UserObjectName>("user_object", "The name of the user object");
+  params.addClassDescription(
+      "Test Interfae User Object postprocessor getting value from InterfaceUO");
```
Interfae -> Interface

vincentlaboure commented on test/src/userobjects/InterfaceUO.C:
```diff
@@ -0,0 +1,75 @@
+validParams<InterfaceUO>()
+{
+  InputParameters params = validParams<InterfaceAverageUserObject>();
+  params.addRequiredCoupledVar("var", "the variable name");
```
the variable name -> The variable name

vincentlaboure commented on test/src/userobjects/InterfaceUO.C:
```diff
@@ -0,0 +1,75 @@
+{
+  InputParameters params = validParams<InterfaceAverageUserObject>();
+  params.addRequiredCoupledVar("var", "the variable name");
+  params.addCoupledVar("var_neighbor", "the variable name");
```
the variable name -> The neighbor variable name

vincentlaboure commented on test/src/userobjects/InterfaceUO.C:
```diff
@@ -0,0 +1,75 @@
+  params.addRequiredCoupledVar("var", "the variable name");
+  params.addCoupledVar("var_neighbor", "the variable name");
+  params.addClassDescription(
+      "Test Interfae User Object computing the average value of two variables across an interface");
```
Interfae -> Interface

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+validParams<InterfaceUO_QP>()
+{
+  InputParameters params = validParams<InterfaceAverageUserObject>();
+  params.addRequiredCoupledVar("var", "the variable name");
```
lines 19-21: same comments as above

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+void
+InterfaceUO_QP::initialize()
+{
+
```
remove empty line

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+InterfaceUO_QP::initialize()
+{
+
+  // define the boundary map nad retrieve element side and boundary_ID
```
nad -> and

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> elem_side_bid =
+      _mesh.buildSideList();
+
+  // retrieve on which boudnary this UO operates
```
boudnary -> boundary

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+  // initialize the map_values looping over all the element and sides
+  for (unsigned int i = 0; i < elem_side_bid.size(); i++)
+  {
+    // check if this boundary
```
missing word in that comment?

arovinelli replied:
reworded

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+    // if this element side is part of the boundary then add elements to the map
+    if (boundaryList.find(std::get<2>(elem_side_bid[i])) != boundaryList.end())
+    {
+
```
remove empty line

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+void
+InterfaceUO_QP::execute()
+{
+
```
remove empty line

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+  auto it = _map_values.find(std::make_pair(_current_elem->id(), _current_side));
+  if (it != _map_values.end())
+  {
+
```
remove empty line

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+    // loop over qps and do stuff
+    for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+    {
+      // vec[qp].resize(1, 0);
```
remove lines 87 and 88

vincentlaboure commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+  else
+    mooseError("InterfaceUO_QP:: cannot fine the required element and side");
```
fine -> find

vincentlaboure reviewed (commented):
@arovinelli Here are some comments, mostly regarding the form rather than the content. I will let the MOOSE guys review that aspect.

Commit:
fixing comments and removing _JxW_neighbor

arovinelli commented:
@permcody and @vincentlaboure 
I'm sorry for the delay, I was at a conference and I didn't check this PR.
I addressed all the comments

moosebuild commented:
Job [App tests](https://civet.inl.gov/job/305483/) on 5f5281a : invalidated by @lindsayad

lindsayad commented on test/doc/content/source/userobjects/InterfaceUO_QP.md:
```diff
@@ -0,0 +1,16 @@
+## Description
+!syntax description /UserObjects/InterfaceUO_QP
+
+InterfaceUO_QP is a user object computing and storing average values across an interface for each quadrature. The kind of average value are the one available in [InterfaceAverageTools](/InterfaceAverageTools.md).
```
Whatever you change the name to for `InterfaceUO` be sure to update this class name. E.g. if `InterfaceUO` becomes `InterfaceAverageUO` then this should become `InterfaceAverageUO_QP`. Which leads me to my primary concern about this PR...we seem to have a ton of objects that are named or do similar things. We have `InterfaceUO` which from its description takes an average, `InterfaceUserObject` which is the base class, and `InterfaceAverageUserObject` which I would think immediately from looking at its name should do something very similar to what `InterfaceUO` is purported to do. When I look at `InterfaceAverageUserObject` I see that most of what it does is to call through to the free-standing functions in `InterfaceAverageTools`. Why can't we have the functionality that's in `InterfaceAverageTools` in the `InterfaceAverageUserObject` class? I'm guessing it's because we also want to use that functionality in the average postprocessor classes... Perhaps we could create an independent object that both the `InterfaceAverageUserObject` and average postprocessor classes inherit from? That would probably be the same amount of code though...so maybe I'm fine with just having that capability in free-standing functions.

So I guess my primary concern is the preponderance of names that appear very similar.

arovinelli replied:
@lindsayad thanks for review. I agree that objects' name are a bit confusing and I should something about it. Before making modifications I would like to ask you a question.
As you may have noticed the `InterfaceUO`, `InterfaceUO_QP`, `InterfaceUserObjectQpAux` and `InterfaceUOPPS`, right now are registered in the `MooseTestApp`. is this ok or you prefer to have them available in the framework? If we let them stay the test app we might rename them from `InterfaceUO`->`InterfaceAverageUO_test`, if we put them in the framework we might do as you suggested `InterfaceUO`->`InterfaceAverageUO`.

Also the `InterfaceAverageTools` is a separate functionality because, as you mentioned` we want it to be available in the postprcessor too. Beside saving code with this structure we might add functionality to both the UO and the PP at once, without the risk of making mistake twice.

I also need to fix the `InterfaceIntegralPostprocessor` because as you said it's not computing an integral however this portion of the code is a bit confusing. I really need to understand how the `getValue`, `threadJoin` works, because it seems to me that when calling getValue bypasses `threadJoin`

lindsayad replied:
> if we put them in the framework we might do as you suggested `InterfaceUO`->`InterfaceAverageUO`.

I'm in favor of trying to condense objects as long as we are not limiting their general utility. I think the core problem is that for someone who is coming in and looking at this code for the first time, it takes a while to realize that there are two different meanings of "average" here. For someone who is used to the MOOSE system of postprocessors, "average" to them likely means a spatial average over the relevant geometric object (e.g. element or side). But you've introduced a second meaning of "average" here. That second meaning is averaging slave and master values. I think we need to try to clearly delineate these two different meanings.

I think that what the objects `InterfaceUO`, `InterfaceUO_QP`, `InterfaceUserObjectQpAux` do are generally applicable and can be included in the framework. Let's just work on naming and clarity of meaning.

arovinelli replied:
@lindsayad if the term average in moose generally associated with a spatial average, In that sense here the only object doing a spatial average is the postprocessor. At an interface we might want an average, a jump, or either something else (basically all the things inside `InterfaceAverageTools`). As you said this is a master slave calculation across the interface, that is not necessarly an average. I need to think about a better name. Let's start from the tools:
what about renaming:

- `InterfaceAverageTolls` -> `InterfaceValueTools`
- `InterfaceUO` -> `InterfaceValueAverageUO` (this is actually computing an average on the Interface)
- `InterfaceUO_QP` -> `InterfaceValueUO_QP`
- `InterfaceUserObjectQpAux` -> `InterfaceValueUO_QP_Aux`

FOR postprocesso the situation is different because they are in fact doing averages of the value. To keep the name short we could do
- `InterfaceIntegralVariableAveragePostprocessor` -> `InterfaceIntegralVariableValuePostprocessor`
- `InterfaceUOPPS` -> `InterfaceValueUOPPS`

waht do you think

lindsayad replied:
I like these names much better

lindsayad reviewed (changes requested):
This is looking really great. See my comment about seemingly redundant names/functionality for my primary reservation about this PR.

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/306083/) on 5f5281a wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

arovinelli commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1340,8 +1341,8 @@ class FEProblemBase : public SubProblem, public Restartable
-   * internal sides.  There are four situations where this may be the case: 1) DGKernels
-   * 2) IntegratedBCs 3)InternalSideUserObjects 4)ElementalAuxBCs
+   * internal sides.  There are five situations where this may be the case: 1) DGKernels
+   * 2) IntegratedBCs 3)InternalSideUserObjects 4)ElementalAuxBCs 5)InternalSideUserObjects
```
sorry, fixed

Commit:
simplified the postprocessor and renamed objectes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/306594/) on dc235cf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/12902/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/12902/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7e388a036e7aa240c09e132d646c3ef6191ad89f`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/306652/) on f562596 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/12902/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/12902/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7e388a036e7aa240c09e132d646c3ef6191ad89f`

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/306736/) on ffc062b wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

arovinelli commented:
@lindsayad for what concern UO we should be fine for the naming convention.
For Postprocessor, you might still want to check it and tell me if you want to change names to the object... an independent opinion would be nice. BTW I took a post processor object by including the `InterfaceValueTools` in the base class. 
Thanks

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/306760/) on 4d3d062 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

lindsayad commented on test/src/userobjects/InterfaceValueAverageUO.C:
```diff
@@ -0,0 +1,75 @@
+{
+
+  for (unsigned int qp = 0; qp < _q_point.size(); ++qp)
+    _value += computeInterfaceValueType(_u[qp], _u_neighbor[qp]) * _JxW[qp];
```
Let's also multiply by `_coord[qp]` as we do for other integrations

lindsayad commented on test/src/postprocessors/InterfaceValueUOPPS.C:
```diff
@@ -0,0 +1,51 @@
+void
+InterfaceValueUOPPS::execute()
+{
+  _value_pp = _uo.getValue();
```
I am very concerned that this will not work correctly. I need to go over our user object and postprocessor execution, but here's what I think could happen: this gets executed after your user object gets executed. The value that you are retrieving from the user object will not have been finalized, e.g. we will not have done the necessary `gatherSums` across processors. So the final value that you have here in this postprocessor will not be equal to the final value in your user object.

A couple of questions:
1. I wanted to quickly verify my hypothesis that this won't quite be right by running one of your test files. I tried running `interface_value_user_object.i` but the solve did nothing; the initial non-linear residual was 0, so nothing ever really happened. Is that test really supposed to do nothing but report a bunch of zeros?
2. Do you really need the `InterfaceValueAverageUO` object or could you not just do all its work in your current `InterfaceValueUOPPS` object since it seems like the former is really just doing a computation that a postprocessor would do, e.g. it's building a single aggregate value?

arovinelli replied:
>  this gets executed after your user object gets executed. The value that you are retrieving from the user object will not have been finalized, e.g. we will not have done the necessary gatherSums

I noticed this behaviour and I don't understand why these has been conceived in this manner. 
This is also confusing when coding in general. Values from UO should always be retrieved after finalize and threadJoin (is there a way to do this?).

> the initial non-linear residual was 0, so nothing ever really happened. Is that test really supposed to do nothing but report a bunch of zeros?
I didn't realize it's not solving anything. wec ould try to see if it woprks properly simply leaving the same problem but changing IC of `u`,`v`, such that the interface kernel has to do something.
However It would be nice to have something more robust like a know solution at the interface for both variables.

> Do you really need the InterfaceValueAverageUO object or could you not just do all its work in your current InterfaceValueUOPPS object since it seems like the former is really just doing a computation that a postprocessor would do, e.g. it's building a single aggregate value?

I agree with you as long as updated finalize postprocessor values can be used inside other objects like Materials, Kernels ,etc. If this is the case we can get rid of `InterfaceValueAverageUO`. and rename `InterfaceValueUOPPS` -> `InterfaceValuePPS` and make all the calculations inside it. This would also imply removing some test.

lindsayad replied:
Yes postprocessor values can absolutely be used inside `Materials`, `Kernels`, etc. See `test/tests/src/kernels/DefaultPostprocessorDiffusion.C`. Let's rename `InterfaceValueUOPPS` to `InterfaceAverageValuePostprocessor` because we are computing a spatial average **and** we are leveraging your `InterfaceValue` capabilities. Sound good?

arovinelli replied:
@lindsayad  sounds great! This will also clean a bit the all PR and take away some confuision

arovinelli replied:
@lindsayad I complete removed the `InterfaceValueUOPPS` as the `InterfaceIntegralVariableAveragePostprocessor` was already doing this.
Now it should just be a matter of naming this postprocessor class

Commit:
addressed most of the comment if not all. wait for checks

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/309160/) on 2a1a38f : invalidated by @arovinelli

connection timeout

arovinelli commented:
@lindsayad and @permcody 

I pushed a new version and now I get an error that I'm not familiar with:

`//: cp /home/moosetest/singularity/start_moosebuild.sh /home/moosetest/singularity/.civet_hpcbuild6_start_moosebuild.iKNeY
//: source /home/moosetest/singularity/.civet_hpcbuild6_start_moosebuild.iKNeY
//: mkdir -p /tmp/Ubuntu-16
//: chmod o-rwx /tmp/Ubuntu-16
//: mapping /tmp/Ubuntu-16 as /tmp within container
//: Using an arbitrary package_builder release: 56178942a6f4cf47dfc6bb8b678d26bb6c5a32ed
//: ARCH=Ubuntu-16.04
//: BUILD_DATE=20181203
//: PR_VERSION=https://github.com/idaholab/package_builder/pull/179
//: rm /home/moosetest/singularity/.civet_hpcbuild6_start_moosebuild.iKNeY
//: cp /tmp/tmpnRdc16 /tmp/Ubuntu-16/
//: /opt/singularity/bin/singularity exec --no-home /home/moosetest/singularity/containers/Ubuntu-16.simg /tmp/tmpnRdc16
Loaded modules 'civet/.civet mpich-gcc-petsc_default-vtk advanced_modules autotools cmake'
Date: Mon Mar 25 18:56:48 MDT 2019
Machine: hpcbuild6
Singularity build enabled
Host Machine Distribution:
LSB Version:	core-9.20160110ubuntu0.2-amd64:core-9.20160110ubuntu0.2-noarch:security-9.20160110ubuntu0.2-amd64:security-9.20160110ubuntu0.2-noarch
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial
MOOSE Package: ARCH=Ubuntu-16.04
BUILD_DATE=20181203
PR_VERSION=https://github.com/idaholab/package_builder/pull/179
Build Root: /opt/civet/build_0
Trigger: Pull request
Step: Pre test (1)
MODULESHOME: /opt/moose/Modules/3.2.10
Currently Loaded Modulefiles:
  1) civet/.civet
  2) moose/.gcc-7.3.1
  3) moose/.mpich-3.2_gcc-7.3.1
  4) moose/.petsc-3.9.4_mpich-3.2_gcc-7.3.1-opt
  5) moose/.VTK-7.1.0_gcc-7.3.1
  6) moose/.tbb-2018_U3
  7) moose/.cppunit-1.12.1_gcc-7.3.1
  8) ccache
  9) miniconda
 10) moose-tools
 11) mpich-gcc-petsc_default-vtk
 12) advanced_modules
 13) autotools
 14) cmake
Base commit: git@github.com:idaholab/moose.git:devel:ffe6d70df12a7d87629c662f2e38c3ac2f002b8c
Head commit: git@github.com:arovinelli/moose.git:INTERFACEUO:2a1a38fe8224906c3c9e9598e6fe2f2e6ecc3b5b

BUILD_ROOT/moose/: git reset --hard origin/master
HEAD is now at 42c6b1f0e4 Merge commit '951cc84d7c6252efd19d2f3958ed7a5daa2a41dc'
BUILD_ROOT/moose/: git submodule deinit --force large_media
Cleared directory 'large_media'
Submodule 'large_media' (../../idaholab/large_media) unregistered for path 'large_media'
BUILD_ROOT/moose/: git submodule deinit --force libmesh
Cleared directory 'libmesh'
Submodule 'libmesh' (../../libMesh/libmesh) unregistered for path 'libmesh'
BUILD_ROOT/moose/: git submodule deinit --force petsc
Cleared directory 'petsc'
Submodule 'petsc' (https://bitbucket.org/petsc/petsc) unregistered for path 'petsc'
BUILD_ROOT/moose/: find . -type f -name .git -delete
BUILD_ROOT/moose/: git fetch --tags origin
ssh: connect to host github.com port 22: Connection timed out

fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
BUILD_ROOT/moose/: git clean -xffdq
BUILD_ROOT/moose/: git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
BUILD_ROOT/moose/: git reset --hard origin/master
HEAD is now at 42c6b1f0e4 Merge commit '951cc84d7c6252efd19d2f3958ed7a5daa2a41dc'
BUILD_ROOT/moose/: git branch -D test
Deleted branch test (was 42c6b1f0e4).
BUILD_ROOT/moose/: git checkout devel
Switched to branch 'devel'
Your branch is behind 'origin/devel' by 27 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
BUILD_ROOT/moose/: git reset --hard origin/devel
HEAD is now at c78b8d1f19 Merge pull request #13099 from tophmatthews/ad_eigenstrain_13091
BUILD_ROOT/moose/: git checkout -b test origin/devel
Switched to a new branch 'test'
Branch 'test' set up to track remote branch 'devel' from 'origin'.
BUILD_ROOT/moose/: git reset --hard origin/devel
HEAD is now at c78b8d1f19 Merge pull request #13099 from tophmatthews/ad_eigenstrain_13091
BUILD_ROOT/moose/: git pull git@github.com:arovinelli/moose.git INTERFACEUO
ssh: connect to host github.com port 22: Connection timed out

fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
ERROR: exiting with code 1
//: rm -rf /tmp/Ubuntu-16`

what should I do?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/309160/) on 2a1a38f : invalidated by @arovinelli

wiewred error

lindsayad commented:
This is not your fault. It’s a CIVET issue. We’ll be working on it tomorrow 

> On Mar 25, 2019, at 7:16 PM, MOOSE Build <notifications@github.com> wrote:
> 
> Job Precheck on 2a1a38f : invalidated by @arovinelli
> 
> wiewred error
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/309176/) on 2a1a38f wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/309246/) on dcfc77e wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

lindsayad commented on test/src/userobjects/InterfaceValueUO_QP.C:
```diff
@@ -0,0 +1,97 @@
+  _map_values.clear();
+
+  // initialize the map_values looping over all the element and sides
+  for (unsigned int i = 0; i < elem_side_bid.size(); i++)
```
We are working hard to try and eliminate the use of `unsigned int` when looping over STL containers because the type returned by the `size()` method is `std::size_t`. We have a helper macro that we've recently added that makes it so you never have to think about the type used for your index. Make sure that this is rebased onto the most recent `next` branch pointer, and then you can use this syntax:
```c++
for (MooseIndex(elem_side_bid) i = 0; i < elem_side_bid.size(); i++)
```
Using this design pattern there will never be a mismatch between the type of `i` and the type of `elem_side_bid.size()`. If you want to know about the macro you can search for `MooseIndex` in `MooseTypes.h`

lindsayad commented on test/src/userobjects/InterfaceValueUO_QP.C:
```diff
@@ -0,0 +1,97 @@
+  return params;
+}
+
+InterfaceValueUO_QP::InterfaceValueUO_QP(const InputParameters & parameters)
```
Let's name this `InterfaceQpValueUserObject` as we don't traditionally use underscores in our file names and if we're going to reference the type of the object e.g. UserObject, we traditionally spell it out and put it as the suffix

lindsayad reviewed (changes requested):
Make sure to go to this pull request on github and click on the `Files changed` tab. There are some old comments from me requesting some re-wording in the doxygen. I would like to see those changes. One those are addressed and the three comments I've left here are addressed, I'm ready to press the merge button on this. Great work!

Commit:
renaming PPS

Commit:
Rename of the postprocessor.
This might cause some errors.

Commit:
upl oading correct goldfile for interface_average_variable_value_postprocessor

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/311827/) on 039ed95 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

Commit:
moved InterfaceValueUO_QP to MooseApp

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/311864/) on 2d24ced wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

Commit:
finish reorganizing the PR

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/311897/) on 1e9c6be wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

arovinelli commented:
@lindsayad I did some cleaning, removed a useless test and checked all your comments.
You might want to give this a final look. please Also tell me if we need to squash the commit and or rebase.

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/312022/) on acd7009 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

lindsayad reviewed (changes requested):
Please rename `InterfaceValueUO_QP` to `InterfaceQpValueUserObject`. Some squashing would be nice. Any time there's commit messages like "we should be good, I hope", that usually indicates squashing would be in good order. However, if you run into any conflicts, which can happen during squashing attempts, don't strain yourself, e.g. I would prefer some squashing but would accept this as is if you can't

arovinelli commented:
@lindsay I'll fix these issues after lunch, and thanks for catching them. If you want I can also try a full squash (just having one  commit) if this makes things better for you

Commit:
renamed InterfaceValueUO_QP->InterfaceQpValueUserObject

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/313056/) on 958efe0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/12902/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/12902/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7e388a036e7aa240c09e132d646c3ef6191ad89f`

arovinelli commented:
@lindsay I was able to rebase all the PR in a single commit, do you want me to push it or you prefer having the history of this?

Commit:
fixed doc filename for InterfaceQpValueUserObject

lindsayad commented:
Since this is pretty much all addition, e.g. you are not changing ownership of lines so git blame should still lead to something meaningful, I would prefer to have the history

arovinelli commented:
Let me see if I can do some partial squashing them as you suggested before

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/313247/) on 8194174 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

arovinelli commented:
@lindsay I did i the partial squash as you requested, this version should be ready to be merged. If you don't have other concerns please go ahead and merge

lindsayad reviewed (approved):
I'm really excited to see your contribution go in. Let's hope it passes all tests!

arovinelli commented:
@lindsayad I'm exited too and of course hoping for the best! This has been a long journey and I want to thanks again you all, especially @lindsayad, @YaqiWang and @vincentlaboure! finger crossed!

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/313304/) on 5f09014 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/12902/coverage)

arovinelli commented:
@lindsayad, @YaqiWang, @vincentlaboure, @jiangwen84  finally this is in! Thanks all again, now we need to work #12066 and get al this thing working!
